### PR TITLE
ci: validate every release candidate without path-filter skips

### LINKED ISSUE #141: ci(release): run complete candidate validation without path-filter skips
## Summary

The exact audited SHA has green three-OS jobs but skips nextest, overflow-off, feature and WASM steps on a workflow-only change. This is insufficient evidence for a final release candidate.

Audit finding **Q1**, priority **P2**. Required v0.6.0 release gate.

## Evidence and reproduction

Audited commit: [263ec19e807454eac943681993623989fb986cc6](https://github.com/Limen-Neural/neuromod/commit/263ec19e807454eac943681993623989fb986cc6). [Source location](https://github.com/Limen-Neural/neuromod/blob/263ec19e807454eac943681993623989fb986cc6/.github/workflows/ci.yml#L62).

```text
gh run view 35164739248 --repo Limen-Neural/neuromod --json jobs
cargo test --locked --all-features
cargo test --locked --release --all-features
```

**Observed:** All Linux/macOS/Windows Build & Test jobs report skipped nextest on SHA 263ec19. Linux Codecov ran tests, so this is a release-matrix coverage gap, not a claim that no tests ran.

## Acceptance criteria

- [x] Provide an unconditional, nonpublishing release-candidate validation path on the selected SHA.
- [x] Run unit tests and doctests on Linux/macOS/Windows; nextest alone does not cover doctests.
- [x] Run Linux formatting, strict Clippy/docs, feature powerset, WASM, overflow semantics, current dependency scan, examples, benchmarks smoke and package dry-run.
- [x] Include toolchain/workflow inputs in routine path-filter coverage where applicable.
- [x] Demonstrate complete runs with no required test step silently skipped; pin third-party Actions.

## Relationships

- Parent: #131 (native sub-issue).
- Blocked by: none. Coordinate overlapping files; no artificial ordering dependency is asserted.
- Blocks: #131 (native blocking relationship).
- Related: #94 and #30. [Exact-head CI evidence](https://github.com/Limen-Neural/neuromod/actions/runs/35164739248).
- Open PR collisions: none at filing preflight.

## Scope boundaries

Resolve this finding and its regression coverage only. This issue does not authorize publication, a tag, unrelated API redesign, or declaring the entire release ready. GitHub is canonical; keep the Linear twin aligned.

## Attribution

Audit reproduction and issue drafting: **Codex (OpenAI)**, requested by and assigned to **@rmems**. Audit date: **2026-09-16 (America/Chicago)**. Earlier related work retains its original attribution.

Maintainer-local evidence: `/home/raulmc/audits/neuromod-0.6.0-263ec19/` (not a public URL). The immutable source links and reproduction above are the portable evidence.

## Reopened benchmark execution gap — 2026-09-17 UTC

[PR 148](https://github.com/Limen-Neural/neuromod/pull/148) delivered unconditional candidate checks and three-platform unit tests/doctests, including verified [workflow-only](https://github.com/Limen-Neural/neuromod/actions/runs/35170149913) and [documentation-only incremental](https://github.com/Limen-Neural/neuromod/actions/runs/35170674256) runs. A completion audit found that hosted CI only runs `cargo bench --locked --no-run --all-features`; this compiles benchmarks but does not satisfy this issue's runtime smoke criterion. The gate is reopened for a focused workflow follow-up executing all four Criterion targets with `--test`, with actual hosted execution evidence required before closure. Existing issue/PR relationships and prior evidence are retained. The separate final-candidate audit runner already includes four runtime smoke commands; it has not run against a final candidate. [The v0.7 benchmark-logic follow-up](https://github.com/Limen-Neural/neuromod/issues/145) remains excluded.

Reproduction and reconciliation: **Codex (OpenAI)** for @rmems. No release publication or tag is authorized.

## Verified gap closure — 2026-09-17 UTC

Maintainer-authorized [PR 154](https://github.com/Limen-Neural/neuromod/pull/154) is squash-merged as `4de7931a48dcc80b0854c29c13b9af7cab8abe45`. Exact head `bc01ac9e599f2b36510ea4ca62227d122b872b07` passed checks and independent review with no unresolved threads. [Hosted Linux execution](https://github.com/Limen-Neural/neuromod/actions/runs/35184271362/job/105082852038) ran all four Criterion targets: neuron 10, STDP 15, memory 17, modulation 19; all 61 cases passed. This closes the runtime gap described above. Original twin [LIM-1361](https://linear.app/rpd-34/issue/LIM-1361) is verified Done, preserving the issue and both PR attachments. Final release-candidate archive validation remains under the [parent release tracker](https://github.com/Limen-Neural/neuromod/issues/131).

Integration and verification: **Codex (OpenAI)** for @rmems.

### PR DESCRIPTION
**Pull request by rmems**
## **User description**
Run the full candidate test path for every main push and pull request, including workflow-only and documentation-only changes, and provide manual dispatch. Linux, macOS, and Windows execute unit tests and doctests; Linux also runs release tests, overflow semantics, feature/WASM checks, strict docs, dependency audit, all five examples in debug/release, benchmark compilation, and package verification.

Closes #141
Linear: https://linear.app/rpd-34/issue/LIM-1361

Removed the path filter and conditional test skips. Retained the pinned Rust version and immutable action SHAs. No publishing workflow is added.

Validation: actionlint and configuration review pass; local default/all-feature/release suites each pass 196 unit tests + 13 doctests, with 12 overflow regressions passing. Independent review caught the omitted sparse GIF example; the append-only fix covers all five Cargo example targets, and scoped re-review approved head `0670aae54c58236ee840b89e289a5b33f5d3f200`. Both sparse GIF profiles were executed successfully.

This PR initial workflow-only revision passed every remote check, directly verifying that workflow-only changes no longer skip tests. The latest append-only commit updates the README CI description in response to review; its remote jobs also exercise the documentation-only path. Remote job completion remains a merge gate; final integrated release validation remains tracked by #131.

Implementation, review coordination, and validation: **Codex (OpenAI)** at @rmems's request.

___

## **CodeAnt-AI Description**
Run complete release-candidate validation for every change

### What Changed
- CI now runs the full test and validation workflow for every main-branch push and pull request, including documentation- and workflow-only changes
- Added manual workflow dispatch for on-demand validation
- Linux now checks release tests, feature combinations, browser WASM support, dependency vulnerabilities, generated documentation, all examples in debug and release modes, benchmark compilation, and package creation
- Added release-mode validation for the standalone crates.io example
- Validation failures now report explicit errors for forbidden documentation terms, incorrect feature activation, and disallowed example dependencies

### Impact
`✅ Fewer untested workflow and documentation changes`
`✅ Earlier detection of release and packaging failures`
`✅ On-demand release-candidate validation`
<details><summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

---
## Summary by cubic
Removes the path filter that skipped CI for workflow-only and documentation-only changes, so every push and pull request to main now runs the full validation suite. Also adds manual dispatch, new Linux checks, and updates the README's CI documentation to match.

- Replaces `!`-based checks with explicit error handling so failures no longer pass silently.
- Switches all OSes from `cargo-nextest` to `cargo test --locked --all-features` for unit tests and doctests.
- Runs release tests, overflow semantics, feature matrix, WASM checks, dependency audit, example smoke tests, benchmark compilation, and package dry run on Linux, plus `cargo run --release` in the crates.io demo.

Closes #141.

<sup>Written for commit 0670aae54c58236ee840b89e289a5b33f5d3f200. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/Limen-Neural/neuromod/pull/148?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -82,19 +70,20 @@ jobs:
...
-        if: steps.changes.outputs.rust == 'true'
-        shell: bash
-        run: |
-          set -euo pipefail
-          cargo install cargo-nextest --locked
-          cargo nextest run --all-features --no-fail-fast
+      # `cargo test` runs both unit tests and doctests. Keep it unconditional:
+      # workflow, toolchain, and documentation-only changes are release inputs.
+      - name: Test (unit tests and doctests)
+        run: cargo test --locked --all-features
```
[the remark this answers is not in the document]
Comment (rmems) [reply .github/workflows/ci.yml:76]: [from Codex]: Updated the README Core CI section in 0670aae54c58236ee840b89e289a5b33f5d3f200 to match the unconditional three-OS unit/doctest workflow, manual dispatch, and Linux release guards. Independent scoped review confirmed the prose matches ci.yml; the separate Codecov nextest description remains accurate.
